### DISCUSSION TITLE
Git Hooks

### DISCUSSION THREAD

**Discussion by chrisvander** [jdx/mise#1764] [Ideas]
One thing that I commonly set up on repositories that could easily be replaced Mise would be Git Hooks. Currently, I have to install Husky and do an additional setup step which is commonly in the `package.json` "prepare" script. I could envision a `[hooks.commit-msg]` block with a similar configuration to the task runner. Or, could reference a task directly. Those hooks could be set up when `mise install` is run inside of a fresh repository. What do you think?

And if hooks don't make sense for the roadmap, something that would make this flow easier is a `prepare` or `postinstall` task that runs immediately after `mise install`.

**Comment by syhol** [comment]
Couldn't you do this with a source'd script in the config file? https://mise.jdx.dev/environments.html#env-source

**Reply by chrisvander** [reply]
That sources the script any time the directory is loaded, and isn't necessarily in sync with lifecycle events.

**Comment by JamBalaya56562** [comment]
This exists, though as a generator rather than a config section.

```console
mise generate git-pre-commit --write --task pre-commit
```

writes `.git/hooks/pre-commit` and makes it executable, so `git commit` runs `mise run pre-commit`. That replaces the Husky install step and the `package.json` `prepare` script — the hook body is just a mise task.

The name undersells it: **`--hook <HOOK>`** picks the destination, defaulting to `pre-commit`. So the `[hooks.commit-msg]` from the original post is:

```console
mise generate git-pre-commit --write --hook commit-msg --task lint-commit-msg
```

The generated body is the same for every hook:

```sh
STAGED="$(git diff-index --cached --name-only -z HEAD | xargs -0)"
export MISE_PRE_COMMIT=1
exec mise run <task> "$@"
```

That trailing `"$@"` is what lets a `commit-msg` task see the commit-message path — it is appended to the task's command line, so `run = 'commitlint --edit'` becomes `commitlint --edit .git/COMMIT_EDITMSG`. It landed in #11801 and is **not yet in a release**; on v2026.8.3 the arguments are still dropped, so a `commit-msg` task cannot see its file yet.

Two things to know regardless of hook: `MISE_PRE_COMMIT=1` is set for all of them, and `STAGED` is a pre-commit concept computed against `HEAD` — so the generated hook errors on a repository's very first commit.

To be clear about the naming collision: mise's own `[hooks]` section is unrelated — it covers `enter`, `leave`, `cd`, `preinstall` and `postinstall`, i.e. mise's lifecycle, not git's. @syhol's `_.source` suggestion runs on directory load, which is the mismatch @chrisvander already identified.
